python/ql/src/experimental/semmle/python/frameworks/Stdlib.qll

```diff
@@ -9,3 +9,84 @@ private import semmle.python.dataflow.new.TaintTracking
 private import semmle.python.dataflow.new.RemoteFlowSources
 private import experimental.semmle.python.Concepts
 private import semmle.python.ApiGraphs
+
+private module NoSQL {
+  private class PyMongoMethods extends string {
+    // These are all find-keyword relevant PyMongo collection level operation methods
+    PyMongoMethods() {
+      this in [
+          "find", "find_raw_batches", "find_one", "find_one_and_delete", "find_and_modify",
+          "find_one_and_replace", "find_one_and_update"
+        ]
+    }
+  }
+
+  private class PyMongoClientCall extends DataFlow::CallCfgNode, NoSQLQuery::Range {
+    PyMongoClientCall() {
+      this =
+        API::moduleImport("pymongo")
+            .getMember("MongoClient")
+            .getReturn()
+            .getAMember*()
+            .getMember(any(PyMongoMethods pyMongoMethod))
+            .getACall()
```

Comment:
I'm a bit skeptical about the `.getAMember*()` bit. Do you really need to be able to handle anything besides `client.<db>.<collection>`? If not, I would suggest writing it like this instead:

```suggestion
        API::moduleImport("pymongo")
            .getMember("MongoClient")
            .getReturn()
            .getAMember() // db, which can have any name
            .getAMember() // collection, which can have any name
            .getMember(any(PyMongoMethods pyMongoMethod))
            .getACall()
```

and similarly for the other case of `getAMember*()` below :)

Reply:
Let me research this before I make this change.

Reply:
This is the correct. I'll make the change when I add in the TypeTracker code.

Reply:
Just for reference: I dug into looking at cases where a developer could call `client.db.list_collection_names()` and if it returned an array (which is does), could a pymongo sink method be called and trigger a false-positive. The answer is no. There are no built-in methods with the same name as the pymongo sink methods available to a Python list nor a Python dictionary.